### PR TITLE
[CARBONDATA-2917] Support binary datatype

### PR DESCRIPTION
**Pull request by QiangCai**
Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6458/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8144/

File: processing/src/main/java/org/apache/carbondata/processing/loading/converter/impl/BinaryFieldConverterImpl.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  public BinaryFieldConverterImpl(DataField dataField, String nullformat, int index,
+      boolean isEmptyBadRecord) {
+    this.dataType = dataField.getColumn().getDataType();
+    this.measure = (CarbonMeasure) dataField.getColumn();
+    this.nullformat = nullformat;
+    this.index = index;
+    this.isEmptyBadRecord = isEmptyBadRecord;
+    this.dataField = dataField;
+  }
+  @Override public void convert(CarbonRow row, BadRecordLogHolder logHolder)
```
Comment (xuchuanyin) [inline processing/src/main/java/org/apache/carbondata/processing/loading/converter/impl/BinaryFieldConverterImpl.java]: why not call the next method internally?

File: processing/src/main/java/org/apache/carbondata/processing/loading/converter/impl/BinaryFieldConverterImpl.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+          LOGGER.debug("Can not convert value to Binary type value. Value considered as null.");
+        }
+        logHolder.setReason(
+            CarbonDataProcessorUtil.prepareFailureReason(measure.getColName(), dataType));
+        row.update(null, index);
+      }
+    }
+  }
+
+  @Override public Object convert(Object value, BadRecordLogHolder logHolder)
```
Comment (xuchuanyin) [inline processing/src/main/java/org/apache/carbondata/processing/loading/converter/impl/BinaryFieldConverterImpl.java]: Move 'Override' to the previous line

File: processing/src/main/java/org/apache/carbondata/processing/loading/converter/impl/BinaryFieldConverterImpl.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    this.index = index;
+    this.isEmptyBadRecord = isEmptyBadRecord;
+    this.dataField = dataField;
+  }
+  @Override public void convert(CarbonRow row, BadRecordLogHolder logHolder)
+      throws CarbonDataLoadingException {
+    String value = row.getString(index);
+    if (value != null) {
+      try {
+        byte[] bytes = Hex.decodeHex(value.toCharArray());
```
Comment (xuchuanyin) [inline processing/src/main/java/org/apache/carbondata/processing/loading/converter/impl/BinaryFieldConverterImpl.java]: So it seems the 'Binary' datatype is actually 'HEX' encoded, not 0101..

**Comment by xuchuanyin** [comment]
Problems that may be ignored are that during loading
1. We use a buffer to store one row and the row is 2MB fow now
2. For a column page, we compress it as a byte array and its length is of 'INT' datatype, which means that a page at most 2GB.
So do we have any restrictions on the binary data type?

File: processing/src/main/java/org/apache/carbondata/processing/loading/converter/impl/BinaryFieldConverterImpl.java (hunk 3 above)
Comment (QiangCai) [reply processing/src/main/java/org/apache/carbondata/processing/loading/converter/impl/BinaryFieldConverterImpl.java]: Hex string is smaller than binary string

**Comment by QiangCai** [comment]
@xuchuanyin 
Yes, better to check the limitation. 
What's your opinion about how to fix it?

**Comment by xuchuanyin** [comment]
@QiangCai I'm not sure about how to solve this. Maybe you can try origin spark and impala and observe how they solve this.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/13/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/228/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.3/8467/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/397/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/266/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/435/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/713/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/535/

**Pull request closed without merging**